osf/models/collection.py

```diff
@@ -0,0 +1,153 @@
+from include import IncludeManager
+
+from osf.models.base import BaseModel, GuidMixin
+from osf.models.validators import validate_title
+from website.exceptions import NodeStateError
+from website.util import api_v2_url
+
+class CollectedGuidMetadata(BaseModel):
+    primary_identifier_name = 'guid___id'
+
+    class Meta:
+        order_with_respect_to = 'collection'
+        unique_together = ('collection', 'guid')
+
+    collection = models.ForeignKey('Collection', on_delete=models.CASCADE)
+    guid = models.ForeignKey('Guid', on_delete=models.CASCADE)
+    creator = models.ForeignKey('OSFUser')
+    collected_type = models.CharField(blank=True, max_length=31)
+    status = models.CharField(blank=True, max_length=31)
+
+    @cached_property
+    def _id(self):
+        return self.guid._id
+
+class Collection(GuidMixin, BaseModel):
+    objects = IncludeManager()
+
+    provider = models.ForeignKey('AbstractProvider', blank=True, null=True, on_delete=models.CASCADE)
+    creator = models.ForeignKey('OSFUser')
+    guid_links = models.ManyToManyField('Guid', through=CollectedGuidMetadata, related_name='collections')
+    collected_types = models.ManyToManyField(
+        'contenttypes.ContentType',
+        related_name='+',
+        limit_choices_to={
+            'model__in': ['abstractnode', 'basefilenode', 'collection', 'preprintservice']
+        })
+    title = models.CharField(max_length=200, validators=[validate_title])
+    collected_type_choices = ArrayField(models.CharField(max_length=31), blank=True, default=list)
+    status_choices = ArrayField(models.CharField(max_length=31), blank=True, default=list)
+    is_public = models.BooleanField(default=False, db_index=True)
+    is_promoted = models.BooleanField(default=False, db_index=True)
+    is_bookmark_collection = models.BooleanField(default=False, db_index=True)
+    is_deleted = models.BooleanField(default=False, db_index=True)
+
+    def __unicode__(self):
+        return '{self.title!r}, with guid {self._id!r}'.format(self=self)
+
+    @property
+    def url(self):
+        return '/{}/'.format(self._id)
+
+    def get_absolute_url(self):
+        return self.absolute_api_v2_url
+
+    @property
+    def absolute_api_v2_url(self):
+        return api_v2_url('/collections{}'.format(self.url))
+
+    @property
+    def linked_nodes_self_url(self):
+        return self.absolute_api_v2_url + 'relationships/linked_nodes/'
+
+    @property
+    def linked_registrations_self_url(self):
+        return self.absolute_api_v2_url + 'relationships/linked_registrations/'
+
+    @property
+    def linked_nodes_related_url(self):
+        return self.absolute_api_v2_url + 'linked_nodes/'
+
+    @property
+    def linked_registrations_related_url(self):
+        return self.absolute_api_v2_url + 'linked_registrations/'
+
+    def save(self, *args, **kwargs):
+        first_save = not self.id
+        if self.is_bookmark_collection:
+            if self.title != 'Bookmarks':
```

Comment:
Doesn't need to change, but is there any advantage to having this check over just doing it? More just a theory question.

Reply:
The short answer is that any possible gain from preferring one way over the other is so small that it's not worth the time thinking about it.

The longer answer involves learning more about [python memory management](https://docs.python.org/2/c-api/memory.html#) and [assignment](https://docs.python.org/2/reference/simple_stmts.html#index-3) than you want to.

Reply:
Kinda what I figured that would be the answer 😆